### PR TITLE
feat: support AsyncAPI spec v3

### LINKED ISSUE #168: AsyncAPI spec v3 support in Optimizer
## Problem

Our library is not yet compatible with AsyncAPI v3. This means that users of our library who want to use AsyncAPI v3 will be unable to do so.

## Solution

We will make the necessary updates to our parser and implementation to support AsyncAPI v3. This involves three main tasks:

- [ ] #173
- [x] #172
- [ ] #192

Once https://github.com/asyncapi/parser-js/issues/780 is solved, update parser dependency (might be automated)

## Rabbit Holes

Potential challenges could include unforeseen issues with the parser update.

## Out of bounds

We will not be adding any new features or making any other major changes to the library as part of this bet.

### LINKED ISSUE #173: Update implementation to handle v3 structure
After we have #172  we need to think about what things will change in the implementation of the optimiser to support v3 of the spec.

Related to #168

### LINKED ISSUE #192: use the parser all[component_type] functions instead of parsing the AsyncAPI file in optimizer

### PR DESCRIPTION
**Pull request by KhudaDad414**
### Description:
* Instead of parsing the components ourselves, it now uses parser-js to parse the components (as it should have from the start )
* Add `debug` module and debugging logs.
* Add a new filter so it doesn't remove components that can be reused.
* Update parser so it now supports v3 of spec.
* Update the example so it reads and writes to a file.

**Note 1**: After merging this PR some components like `traits` won't be optimised since from this point forward we are not going to parse the AsyncAPI file here, the parser is doing it. (Read more here: https://github.com/asyncapi/parser-js/issues/820)

fixes #168, fixes #192, fixes #173

### CODE REVIEW

**Comment by KhudaDad414** [comment]
@derberg @jonaslagoni would appreciate some feedback. 🙏

**Comment by KhudaDad414** [comment]
@derberg pingy pongy. 😆

File: package.json [hunk 1]
```diff
@@ -55,7 +56,9 @@
     "access": "public"
   },
   "dependencies": {
-    "@asyncapi/parser": "^2.1.0",
+    "@asyncapi/parser": "v3.0.0-next-major-spec.2",
```
Comment (derberg) [inline package.json:59]:
```suggestion
    "@asyncapi/parser": "3.0.0-next-major-spec.6",
```

**Review by derberg — COMMENTED** [review]
LGTM, but I guess better to wait with merge before changes are done to enable operations under components.

Also maybe this should go out as 1.0

**Comment by KhudaDad414** [comment]
> Also maybe this should go out as 1.0

There are a few more things that need to change for v1. like supporting all components and changing the Report generation format.

**Comment by derberg** [comment]
@KhudaDad414 so now there are no breaking changes here? all good working with v2 documents?

**Comment by KhudaDad414** [comment]
@derberg yes. it is compatible with v3 and v2.

**Review by derberg — DISMISSED** [review]

**Comment by KhudaDad414** [comment]
@derberg tests were failing. can you approve it again, please?

**Review by derberg — APPROVED** [review]
LGTM

**Pull request merged**